### PR TITLE
Refac/organization standardization

### LINKED ISSUE #583: [REFAC] - Standardise Organisation controllers
As part of the standardisation of the reconciliation loop for Greenhouse controllers, these controllers need to be updated to follow the common recommendation for the Reconciliation. This includes DexController, OrganizationController, RBACController, ServiceProxyController, TeamRoleSeederController. 

It is not required to consolidate these controllers at this point.

## Acceptance Criteria

- [ ] Implement the Interface
- [ ] Use the standardised patch for updating the status instead of the deferred update (if applicable)
- [ ] Convert the tests to use [test.TestSetup](https://github.com/cloudoperators/greenhouse/blob/main/pkg/test/resources.go) (if applicable) including creation of missing resources 
- [ ] Remove dedicated finalizers used on the controllers because they are replaced with standard ones.

### PR DESCRIPTION
**Pull request by gciezkowski-acc**
## Description
I implement new approach to the organization controllers.

## What type of PR is this? (check all applicable)

- [ ] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [X] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

- Related Issue # (issue)
- Closes #583 
- Fixes # (issue)

> Remove if not applicable

## Added tests?

- [X] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for tests # (issue/pr)

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Added to documentation?

- [ ] 📜 README.md
- [ ] 🤝 Documentation pages updated
- [X] 🙅 no documentation needed
- [ ] (if applicable) generated OpenAPI docs for CRD changes

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (no documentation needed)
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: pkg/controllers/organization/organization_controller.go [hunk 1]
```diff
@@ -5,6 +5,7 @@ package organization
 
 import (
 	"context"
+	"github.com/cloudoperators/greenhouse/pkg/lifecycle"
```
Comment (kengou) [inline pkg/controllers/organization/organization_controller.go:8]: nit pic, can you fix `go imports` ?

File: pkg/controllers/organization/teamrole_seeder_controller.go [hunk 2]
```diff
@@ -5,6 +5,7 @@ package organization
 
 import (
 	"context"
+	"github.com/cloudoperators/greenhouse/pkg/lifecycle"
```
Comment (kengou) [inline pkg/controllers/organization/teamrole_seeder_controller.go:8]: nit pic, can you fix `go imports` ?

**Review by kengou — CHANGES_REQUESTED** [review]

File: pkg/controllers/organization/organization_controller.go (hunk 1 above)
Comment (gciezkowski-acc) [reply pkg/controllers/organization/organization_controller.go]: Done

File: pkg/controllers/organization/teamrole_seeder_controller.go (hunk 2 above)
Comment (gciezkowski-acc) [reply pkg/controllers/organization/teamrole_seeder_controller.go]: Done

File: pkg/controllers/organization/dex_controller.go [hunk 3]
```diff
@@ -67,25 +68,33 @@ func (r *DexReconciler) SetupWithManager(name string, mgr ctrl.Manager) error {
...
+func (r *DexReconciler) EnsureCreated(ctx context.Context, object lifecycle.RuntimeObject) (ctrl.Result, lifecycle.ReconcileResult, error) {
+	org, ok := object.(*greenhousesapv1alpha1.Organization)
+	if !ok {
+		return ctrl.Result{}, lifecycle.Failed, errors.Errorf("RuntimeObject has incompatible type.")
 	}
 
 	// Ignore organizations without OIDC configuration.
 	if org.Spec.Authentication == nil || org.Spec.Authentication.OIDCConfig == nil {
-		return ctrl.Result{}, nil
+		return ctrl.Result{}, lifecycle.Pending, nil
```
Comment (abhijith-darshan) [inline pkg/controllers/organization/dex_controller.go:86]: Is `Pending` being returned so that the `Ready` status is not set?

**Comment by abhijith-darshan** [comment]
We should merge all 3 controllers into one since they are acting on the same Custom Resource. 

With current approach there will be conflict with `DexReconciler` and the remaining ones.

if no OIDC config is provided then it will be `Pending` and the remaining will be in `Created` state.

There will be potentially an infinite loop where the competing controllers will put `Ready` and Dex will put it back to `Pending`

You can merge all 3 functionalities into one Organization Reconciler and the logical order can be -
- Functionality in OrganizationReconciler (Namespace, etc…)
- Functionality in Rbac
- Lastly the DexReconciler and it can exit with a success if there is no OIDC config. You can set a condition to reflect this state.

**Comment by uwe-mayer** [comment]
> We should merge all 3 controllers into one since they are acting on the same Custom Resource.
> 
> With current approach there will be conflict with `DexReconciler` and the remaining ones.
> 
> if no OIDC config is provided then it will be `Pending` and the remaining will be in `Created` state.
> 
> There will be potentially an infinite loop where the competing controllers will put `Ready` and Dex will put it back to `Pending`
> 
> You can merge all 3 functionalities into one Organization Reconciler and the logical order can be -
> 
> * Functionality in OrganizationReconciler (Namespace, etc…)
> * Functionality in Rbac
> * Lastly the DexReconciler and it can exit with a success if there is no OIDC config. You can set a condition to reflect this state.
> 
> But since the issue states you don’t need to consolidate at this point we need to bring this up in core dev meeting to check with others

Let's go with a `TODO` and a `Predicate` on the `SetupWithManager` func of the Controller to not reconcile `Dex` controller on `Status` changes.

File: pkg/controllers/organization/dex_controller.go (hunk 3 above)
Comment (gciezkowski-acc) [reply pkg/controllers/organization/dex_controller.go]: Done, I switched to lifecycle.Success in that case.

Comment (gciezkowski-acc) [reply pkg/controllers/organization/dex_controller.go]: I move it to the predicate to filter reconciliation.

**Review by kengou — COMMENTED** [review]
looks good to me

**Review by kengou — APPROVED** [review]

**Pull request merged**